From 8602e1629ac3464964bc691465d97ceb0197a433 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:24:17 -0400
Subject: [tools] perf/trace: Use strarray for ltrace's whence arg

Message-id: <1408462094-14194-89-git-send-email-jolsa@redhat.com>
Patchwork-id: 88068
O-Subject: [PATCH RHEL7.1 BZ1131394 088/325] perf trace: Use strarray for ltrace's whence arg
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit efe6b882cda2a9967a629fa14b6106b6a9a558a2
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Tue Sep 3 16:15:12 2013 -0300
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-trace.c b/tools/perf/builtin-trace.c
index aeb3387..449a9e7 100644
--- a/tools/perf/builtin-trace.c
+++ b/tools/perf/builtin-trace.c
@@ -72,31 +72,6 @@ static size_t syscall_arg__scnprintf_hex(char *bf, size_t size,
 
 #define SCA_HEX syscall_arg__scnprintf_hex
 
-static size_t syscall_arg__scnprintf_whence(char *bf, size_t size,
-         struct syscall_arg *arg)
-{
- int whence = arg->val;
-
- switch (whence) {
-#define P_WHENCE(n) case SEEK_##n: return scnprintf(bf, size, #n)
- P_WHENCE(SET);
- P_WHENCE(CUR);
- P_WHENCE(END);
-#ifdef SEEK_DATA
- P_WHENCE(DATA);
-#endif
-#ifdef SEEK_HOLE
- P_WHENCE(HOLE);
-#endif
-#undef P_WHENCE
- default: break;
- }
-
- return scnprintf(bf, size, "%#x", whence);
-}
-
-#define SCA_WHENCE syscall_arg__scnprintf_whence
-
 static size_t syscall_arg__scnprintf_mmap_prot(char *bf, size_t size,
             struct syscall_arg *arg)
 {
@@ -254,10 +229,20 @@ static size_t syscall_arg__scnprintf_futex_op(char *bf, size_t size, struct sysc
  return printed;
 }
 
+#define SCA_FUTEX_OP  syscall_arg__scnprintf_futex_op
+
 static const char *itimers[] = { "REAL", "VIRTUAL", "PROF", };
 static DEFINE_STRARRAY(itimers);
 
-#define SCA_FUTEX_OP  syscall_arg__scnprintf_futex_op
+static const char *whences[] = { "SET", "CUR", "END",
+#ifdef SEEK_DATA
+"DATA",
+#endif
+#ifdef SEEK_HOLE
+"HOLE",
+#endif
+};
+static DEFINE_STRARRAY(whences);
 
 static size_t syscall_arg__scnprintf_open_flags(char *bf, size_t size,
             struct syscall_arg *arg)
@@ -340,7 +325,8 @@ static struct syscall_fmt {
  { .name     = "ioctl",     .errmsg = true,
    .arg_scnprintf = { [2] = SCA_HEX, /* arg */ }, },
  { .name     = "lseek",     .errmsg = true,
-   .arg_scnprintf = { [2] = SCA_WHENCE, /* whence */ }, },
+   .arg_scnprintf = { [2] = SCA_STRARRAY, /* whence */ },
+   .arg_parm  = { [2] = &strarray__whences, /* whence */ }, },
  { .name     = "lstat",     .errmsg = true, .alias = "newlstat", },
  { .name     = "madvise",    .errmsg = true,
    .arg_scnprintf = { [0] = SCA_HEX,  /* start */
-- 
1.7.1